File: C/Hacks/Hack11/airport.c
/*
 * Name: Anthony Candelas
 * Email: tsallinger2@unl.edu
 * Date: 17Nov2025
 * Lab/Task: Hack11 - Airport utilities
 *
 * Description:
 * Implements Airport creation, formatting, distance and travel-time
 * utilities required by the Hack11 assignment.
 */

#include "airport.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* Helper: duplicate string safely (returns NULL if src NULL). */
static char* strCopy(const char* src) {
  if (!src) return NULL;
  char* s = strdup(src);
  return s;
}

void initAirport(Airport* airport,
                 const char* gpsId,
                 const char* type,
                 const char* name,
                 double latitude,
                 double longitude,
                 int elevationFeet,
                 const char* city,
                 const char* countryAbbrv) {
  if (!airport) return;
  airport->gpsId = strCopy(gpsId);
  airport->type = strCopy(type);
  airport->name = strCopy(name);
  airport->latitude = latitude;
  airport->longitude = longitude;
  airport->elevationFeet = elevationFeet;
  airport->city = strCopy(city);
  airport->countryAbbrv = strCopy(countryAbbrv);
}

Airport* createAirport(const char* gpsId,
                       const char* type,
                       const char* name,
                       double latitude,
                       double longitude,
                       int elevationFeet,
                       const char* city,
                       const char* countryAbbrv) {
  Airport* a = malloc(sizeof(Airport));
  if (!a) return NULL;
  initAirport(a, gpsId, type, name, latitude, longitude, elevationFeet, city, countryAbbrv);
  return a;
}

void freeAirport(Airport* a) {
  if (!a) return;
  free(a->gpsId);
  free(a->type);
  free(a->name);
  free(a->city);
  free(a->countryAbbrv);
  free(a);
}

char* airportToString(const Airport* a) {
  if (!a) return NULL;
  /* produce a string: "GPSID (type): name — city, country [lat, lon] elev ft" */
  const char* gps = a->gpsId ? a->gpsId : "";
  const char* type = a->type ? a->type : "";
  const char* name = a->name ? a->name : "";
  const char* city = a->city ? a->city : "";
  const char* country = a->countryAbbrv ? a->countryAbbrv : "";

  char buf[512];
  int n = snprintf(buf, sizeof(buf), "%s (%s): %s — %s, %s [%.6f, %.6f] %d ft",
                   gps, type, name, city, country, a->latitude, a->longitude, a->elevationFeet);
  if (n < 0) return NULL;
  char* out = malloc((size_t)n + 1);
  if (!out) return NULL;
  memcpy(out, buf, (size_t)n + 1);
  return out;
}

/* Convert degrees to radians. */
static double deg2rad(double deg) {
  return deg * M_PI / 180.0;
}

double getAirDistance(const Airport* origin, const Airport* destination) {
  if (!origin || !destination) return 0.0;
  double phi1 = deg2rad(origin->latitude);
  double phi2 = deg2rad(destination->latitude);
  double deltaLambda = deg2rad(destination->longitude - origin->longitude);
  /* spherical law of cosines */
  double cosval = sin(phi1) * sin(phi2) + cos(phi1) * cos(phi2) * cos(deltaLambda);
  /* clamp to [-1,1] to avoid NaN from rounding */
  if (cosval > 1.0) cosval = 1.0;
  if (cosval < -1.0) cosval = -1.0;
  double central = acos(cosval);
  const double R = 6371.0; /* km */
  return central * R;
}

double getEstimatedTravelTime(const Airport* stops,
                              int size,
                              double aveKmsPerHour,
                              double aveLayoverTimeHrs) {
  if (!stops || size < 1 || aveKmsPerHour <= 0.0) return 0.0;
  if (size == 1) return 0.0; /* no travel needed */
  double totalHours = 0.0;
  for (int i = 0; i < size - 1; ++i) {
    double dist = getAirDistance(&stops[i], &stops[i+1]);
    totalHours += dist / aveKmsPerHour;
  }
  /* add layover times for intermediate stops (size-2 layovers) */
  if (size > 2 && aveLayoverTimeHrs > 0.0) {
    totalHours += (size - 2) * aveLayoverTimeHrs;
  }
  return totalHours;
}

File: Java/Hack13/unl/soc/Airport.java
/*
 * Name: Anthony Candelas
 * Email: tsallinger2@unl.edu
 * Date: 17Nov2025
 * Lab/Task: Hack11 - Airport (Java)
 *
 * Description:
 * Java representation of an Airport record and utility methods
 * to compute air distances and estimated travel times.
 */
package unl.soc;

/**
 * Represents an airport record with fields commonly found in the
 * ICAO database: gpsId, type, name, latitude, longitude, elevation,
 * city and country abbreviation.
 */
public class Airport {
    private String gpsId;
    private String type;
    private String name;
    private double latitude;
    private double longitude;
    private int elevationFeet;
    private String city;
    private String countryAbbrv;

    /**
     * Compatibility constructor used by grader tests: parameter order
     * (gpsId, name, latitude, longitude, type, elevationFeet, city, country).
     */
    public Airport(String gpsId, String name, double latitude, double longitude,
                   String type, int elevationFeet, String city, String countryAbbrv) {
        this.gpsId = gpsId;
        this.type = type;
        this.name = name;
        this.latitude = latitude;
        this.longitude = longitude;
        this.elevationFeet = elevationFeet;
        this.city = city;
        this.countryAbbrv = countryAbbrv;
    }

    /**
     * Initialize this Airport's fields. Caller provides values for all fields.
     */
    public void initAirport(String gpsId, String type, String name,
                            double latitude, double longitude, int elevationFeet,
                            String city, String countryAbbrv) {
        this.gpsId = gpsId;
        this.type = type;
        this.name = name;
        this.latitude = latitude;
        this.longitude = longitude;
        this.elevationFeet = elevationFeet;
        this.city = city;
        this.countryAbbrv = countryAbbrv;
    }

    /* Getters */
    public String getGpsId() { return gpsId; }
    public String getType() { return type; }
    public String getName() { return name; }
    public double getLatitude() { return latitude; }
    public double getLongitude() { return longitude; }
    public int getElevationFeet() { return elevationFeet; }
    public String getCity() { return city; }
    public String getCountryAbbrv() { return countryAbbrv; }

    /**
     * Returns a string representation of the airport.
     *
     * @return string representation of the airport
     */
    @Override
    public String toString() {
        return String.format("%-10s %-14s %-40s %7.2f %7.2f %d %s %s",
                             gpsId == null ? "" : gpsId,
                             type == null ? "" : type,
                             name == null ? "" : name,
                             latitude,
                             longitude,
                             elevationFeet,
                             city == null ? "" : city,
                             countryAbbrv == null ? "" : countryAbbrv);
    }

    /**
     * Convert degrees to radians.
     */
    private static double deg2rad(double deg) {
        return deg * Math.PI / 180.0;
    }

    /**
     * Compute air distance (km) from this airport to {@code other} using the
     * spherical law of cosines. Returns 0.0 for null input.
     */
    public double getAirDistance(Airport other) {
        if (other == null) return 0.0;
        double phi1 = deg2rad(this.latitude);
        double phi2 = deg2rad(other.latitude);
        double deltaLambda = deg2rad(other.longitude - this.longitude);
        double cosval = Math.sin(phi1) * Math.sin(phi2) + Math.cos(phi1) * Math.cos(phi2) * Math.cos(deltaLambda);
        if (cosval > 1.0) cosval = 1.0;
        if (cosval < -1.0) cosval = -1.0;
        double central = Math.acos(cosval);
        final double R = 6371.0; /* Earth's radius in kilometers */
        return central * R;
    }

    /**
     * Compute estimated travel time (hours) across the provided stops
     * using the average speed (km/h) and average layover time (hours).
     * This version accepts a {@link java.util.List} of airports (grader usage).
     */
    public static double getEstimatedTravelTime(java.util.List<Airport> stops, double aveKmsPerHour, double aveLayoverTimeHrs) {
        if (stops == null || stops.size() < 1 || aveKmsPerHour <= 0.0) return 0.0;
        if (stops.size() == 1) return 0.0;
        double totalHours = 0.0;
        for (int i = 0; i < stops.size() - 1; ++i) {
            totalHours += stops.get(i).getAirDistance(stops.get(i+1)) / aveKmsPerHour;
        }
        if (stops.size() > 2 && aveLayoverTimeHrs > 0.0) {
            totalHours += (stops.size() - 2) * aveLayoverTimeHrs;
        }
        return totalHours;
    }
}
